### PR TITLE
Add Vector3::cos_to and related optimizations

### PR DESCRIPTION
**Pull request by ghost**
Adds a method to get the cosine of the angle between two vectors faster than `normalized().dot(b.normalized())`, and uses it where applicable.

Sample output from test project (intel i7-10750H):
~~~
a.normalized().dot(b.normalized()) :
  100 :
    9 usecs
  1000 :
    93 usecs
  10000 : 
    736 usecs
  100000 :
    7 msecs, 929 usecs
  1000000 :
    72 msecs, 735 usecs
  10000000 :
    724 msecs, 601 usecs
  100000000 :
    7 secs, 151 msecs, 508 usecs

a.cos_to(b):
  100 :
    4 usecs
  1000 :
    66 usecs
  10000 :
    328 usecs
  100000 :
    3 msecs, 263 usecs
  1000000 :
    31 msecs, 441 usecs
  10000000 :
    315 msecs, 199 usecs
  100000000 :
    2 secs, 976 msecs, 159 usecs
~~~

[test project](https://github.com/godotengine/godot/files/10474998/cos_to_normdot_comparison.zip)

### CODE REVIEW

File: core/math/vector3.h [hunk 1]
```diff
@@ -290,6 +291,14 @@ Vector3 Vector3::project(const Vector3 &p_to) const {
 	return p_to * (dot(p_to) / p_to.length_squared());
 }
 
+real_t Vector3::cos_to(const Vector3 &p_to) const {
+	if (is_zero_approx() || p_to.is_zero_approx()) {
+		return 0; // for consistency with normalized().dot(p_to.normalized())
```
Comment (Sauermann) [inline core/math/vector3.h:296]: Please follow this guideline for formatting comments https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide

**Review by Sauermann — COMMENTED** [review]
Please check, if this function could also be used in:
- Transform2D::get_skew()
- Node3DEditorViewport::update_transform_gizmo_view()
- Animation::_vector2_track_optimize_key(...) Line 3934

You might also want to have a look at tests/scene/test_primitives.h if there are adjustments possible:
- SUBCASE("Only 6 distinct normals.")
- SUBCASE("[Primitive][Sphere] All normals lie in direction of corresponding point.")

File: core/math/vector3.h (hunk 1 above)
Comment (oenixphay42) [reply core/math/vector3.h]: Done.

**Comment by oenixphay42** [comment]
> Please check, if this function could also be used in:
> 
>     * Transform2D::get_skew()
> 
>     * Node3DEditorViewport::update_transform_gizmo_view()
> 
>     * Animation::_vector2_track_optimize_key(...) Line 3934
> 
> 
> You might also want to have a look at tests/scene/test_primitives.h if there are adjustments possible:
> 
>     * SUBCASE("Only 6 distinct normals.")
> 
>     * SUBCASE("[Primitive][Sphere] All normals lie in direction of corresponding point.")

Updated to use in those places.

**Comment by oenixphay42** [comment]
Fixed the missing tabs and added tests for the edge case.

**Comment by oenixphay42** [comment]
Forgot the extra tests for 2D the first time.

**Comment by oenixphay42** [comment]
Rebased (building godot-cpp test finished locally after that)

**Pull request closed without merging**